source/Core/Drivers/BootLogo.cpp

```diff
@@ -20,60 +21,75 @@ void BootLogo::handleShowingLogo(const uint8_t *ptrLogoArea) {
   } else if (ptrLogoArea[0] == 0xAA) {
     showNewFormat(ptrLogoArea + 1);
   }
+
   OLED::clearScreen();
   OLED::refresh();
 }
 
 void BootLogo::showOldFormat(const uint8_t *ptrLogoArea) {
-
   OLED::drawAreaSwapped(0, 0, 96, 16, (uint8_t *)(ptrLogoArea + 4));
   OLED::refresh();
-
-  // Delay here until button is pressed or its been the amount of seconds set by the user
+  // Delay here with static logo until a button is pressed or its been the amount of seconds set by the user
   delay();
 }
 
 void BootLogo::showNewFormat(const uint8_t *ptrLogoArea) {
-  if (getSettingValue(SettingsOptions::LOGOTime) == 0) {
+  if (getSettingValue(SettingsOptions::LOGOTime) == logoMode_t::SKIP) {
     return;
   }
+
   // New logo format (a) fixes long standing byte swap quirk and (b) supports animation
   uint8_t interFrameDelay = ptrLogoArea[0];
   OLED::clearScreen();
-  ButtonState buttons = getButtonState();
 
   // Now draw in the frames
   int position = 1;
-  do {
-
+  while (getButtonState() == BUTTON_NONE) {
     int len = (showNewFrame(ptrLogoArea + position));
     OLED::refresh();
     position += len;
-    buttons = getButtonState();
 
     if (interFrameDelay) {
       osDelay(interFrameDelay * 4);
     }
+
     // 1024 less the header type byte and the inter-frame-delay
-    if (getSettingValue(SettingsOptions::LOGOTime) > 0 && (position >= 1022 || len == 0)) {
-      // Delay here until button is pressed or its been the amount of seconds set by the user
-      delay();
-      return;
+    if (getSettingValue(SettingsOptions::LOGOTime) && (position >= 1022 || len == 0)) {
+      // Animated logo stops here ...
+      if (getSettingValue(SettingsOptions::LOGOTime) == logoMode_t::INFINITY) {
+        // ... but if it's infinite logo setting then keep it rolling over again until a button is pressed
+        osDelay(4 * TICKS_100MS);
+        OLED::clearScreen();
+        position = 1;
+        continue;
+      }
+    } else {
+      // Animation in progress so jumping to the next frame
+      continue;
     }
-  } while (buttons == BUTTON_NONE);
+
+    // Static logo case ends up right here, so delay until a button is pressed or its been the amount of seconds set by the user
+    delay();
+    return;
```

Comment:
TL;DR: I'm not good at binary formats so my understanding of the way how animated logo is processed here is based more like on _intuition_ & tests on hardware.

Just would like to get confirmation on the accuracy of the added **comments** because this block bothers me a bit. I don't think that I fully understand how it works because for me `showNewFrame` lacks of information about constants & values. I've read this [docs](https://github.com/Ralim/IronOS-Meta/tree/main/Bootup%20Logos) and based on this:

_the next byte indicates the frame timing in milliseconds, or **0 to indicate: Only show the first frame for the whole bootloader duration (still image mode)**.

even tried something like:
```
if (ptrLogoArea[1]) {
  deal with animation only in this case
}
```
but I couldn't make it work properly for some reason.

Reply:
Yeah so the logic here is that we have a frame + duration during animations or a 0 to mark it as freezing.
We also have the start markers at the beginning of the area to decide if we decode as the new format or the old format.

I'll have to re-read more in depth but at a skim this looks about right logic-wise for the comments